canonical/vanilla-framework
# Overhaul theming

## Done

Change the way themes are applied from light and dark mixins to CSS variables.

### The issue now
The problem now is that colours are separated from other styles. All set in a different mixin at the end.
If you want to expand or change the component you have to find the `// Theming` section at the end to figure out on which selector those colours are applied: 
```scss
  // Theming
  @if ($theme-default-p-contextual-menu == 'dark') {
    @include vf-contextual-menu-dark-theme;

    [class*='p-contextual-menu'].is-light {
      @include vf-contextual-menu-light-theme;
    }
  } @else {
    @include vf-contextual-menu-light-theme;

    [class*='p-contextual-menu'].is-dark {
      @include vf-contextual-menu-dark-theme;
    }
  }
```

Then, looking at those light and dark mixins you can see that it duplicates the whole selector tree to apply colours where needed

```scss
@mixin vf-contextual-menu-theme(
  // color of the contextual menu background
  $color-contextual-menu-background,
  // color of the contextual menu border between groups
  $color-contextual-menu-border,
  // color of the contextual menu item link text
  $color-contextual-menu-text,
  // color of the contextual menu item link active background
  $color-contextual-menu-active-background,
  // color of the contextual menu item link hover background
  $color-contextual-menu-hover-background
) {
  .p-contextual-menu__dropdown {
    background: $color-contextual-menu-background;
  }

  .p-contextual-menu__group + .p-contextual-menu__group {
    border-top-color: $color-contextual-menu-border;
  }

  .p-contextual-menu__link {
    &,
    &:active,
    &:hover,
    &:visited {
      background-color: transparent;
      color: $color-contextual-menu-text;
    }

    &:hover {
      background-color: $color-contextual-menu-hover-background;
    }

    &:active {
      background-color: $color-contextual-menu-active-background;
      cursor: default;
    }
  }
}
```

This, in my opinion, is very fragile as it separates styles that should be tightly coupled and could lead to inconsistencies very quickly (and I find it a bit hard to work with to work out on which selector the mixin is applied to then from here, try and target the correct one). Plus it can get a little crazy on components with colour variations like chips and buttons.


### The proposed solution

Instead, what I propose is setting the colours at the top of the mixin using css variables. That way we can set the colours alongside every other property like we would do in any project.

First, we set the placeholders for the component
```scss
  %light-version {
    --color-contextual-menu-background: #{$colors--light-theme--background-default};
    --color-contextual-menu-hover-background: #{$colors--light-theme--background-hover};
    --color-contextual-menu-active-background: #{$colors--light-theme--background-active};
    --color-contextual-menu-border: #{$colors--light-theme--border-default};
    --color-contextual-menu-text: #{$colors--light-theme--text-default};
  }

  %dark-version {
    --color-contextual-menu-background: #{$colors--dark-theme--background-default};
    --color-contextual-menu-hover-background: #{$colors--dark-theme--background-hover};
    --color-contextual-menu-active-background: #{$colors--dark-theme--background-active};
    --color-contextual-menu-border: #{$colors--dark-theme--border-default};
    --color-contextual-menu-text: #{$colors--dark-theme--text-default};
  }
```

Then in the main selector/placeholder, we extend those based on the theme
```scss
    @if ($theme-default-p-contextual-menu == 'dark') {
      @extend %dark-version;
    } @else {
      @extend %light-version;
    }

    &.is-light {
      @extend %light-version;
    }

    &.is-dark {
      @extend %dark-version;
    }
```

And then we can just declare the colours alongside the other properties using the variables
```scss
  .p-contextual-menu__dropdown {
    @extend %vf-has-box-shadow;
    @extend %vf-has-round-corners;

    background: var(--color-contextual-menu-background);
    display: none;
    margin: 0;
    ...
```

### Going further

I kept the default theming logic 
```scss
    @if ($theme-default-p-contextual-menu == 'dark') {
      @extend %dark-version;
    } @else {
      @extend %light-version;
    }
``` 
as to not create a breaking change.

This is controlled by the https://github.com/canonical/vanilla-framework/blob/main/scss/_settings_themes.scss#L1-L10 file
```scss
$theme-default-forms: 'light' !default;
$theme-default-hr: 'light' !default;
$theme-default-nav: 'light' !default;
$theme-default-p-side-navigation: 'light' !default;
$theme-default-p-search-box: 'light' !default;
$theme-default-p-divider: 'light' !default;
$theme-default-p-contextual-menu: 'light' !default;
$theme-default-p-inline-list--middot: 'light' !default;
$theme-default-p-button: 'light' !default;
$theme-default-p-chip: 'light' !default;
```
I personally think this is a useless abstraction and not the way we should think about themes.
If we removed it, the way themes are applied could be simplified even further by getting rid of the placeholders:

```scss
  %p-contextual-menu {
    display: inline-block;
    margin: 0;
    position: relative;

    --color-contextual-menu-background: #{$colors--light-theme--background-default};
    --color-contextual-menu-hover-background: #{$colors--light-theme--background-hover};
    --color-contextual-menu-active-background: #{$colors--light-theme--background-active};
    --color-contextual-menu-border: #{$colors--light-theme--border-default};
    --color-contextual-menu-text: #{$colors--light-theme--text-default};

    &.is-dark {
        --color-contextual-menu-background: #{$colors--dark-theme--background-default};
        --color-contextual-menu-hover-background: #{$colors--dark-theme--background-hover};
        --color-contextual-menu-active-background: #{$colors--dark-theme--background-active};
        --color-contextual-menu-border: #{$colors--dark-theme--border-default};
        --color-contextual-menu-text: #{$colors--dark-theme--text-default};
    }
  }
```

And then going even further, if we imagine that the theme is applied globally by a class on the `<body>` element, we could have all the colour settings abstracted into this and away from the components themselves.


## QA

- Open [demo](https://vanilla-framework-4771.demos.haus/)
- Check that all the updated components look the same
- Check Percy to make sure nothing else broke.

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
  - if CSS class names are not changed it can be bugfix relesase (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
